conda/core/solve.py

```diff
@@ -417,23 +479,37 @@ def _add_specs(self, ssc):
                 target_prec = matches_for_spec[0]
                 if target_prec.is_unmanageable:
                     ssc.specs_map[pkg_name] = target_prec.to_match_spec()
-                elif ssc.update_modifier == UpdateModifier.FREEZE_INSTALLED:
-                    if pkg_name in context.aggressive_update_packages:
-                        ssc.specs_map[pkg_name] = pkg_name
-                    elif target_prec in explicit_spec_package_pool:
-                        ssc.specs_map[pkg_name] = target_prec.to_match_spec()
-                    elif pkg_name in ssc.specs_from_history_map:
-                        ssc.specs_map[pkg_name] = ssc.specs_from_history_map[pkg_name]
+                elif MatchSpec(pkg_name) in context.aggressive_update_packages:
+                    ssc.specs_map[pkg_name] = pkg_name
+                elif self._should_freeze(ssc, target_prec, conflict_specs, explicit_pool,
+                                         installed_pool):
+                    ssc.specs_map[pkg_name] = target_prec.to_match_spec()
+                elif pkg_name in ssc.specs_from_history_map:
+                    ssc.specs_map[pkg_name] = ssc.specs_from_history_map[pkg_name]
                 else:
-                    ssc.specs_map[pkg_name] = MatchSpec(spec, target=target_prec.dist_str())
+                    ssc.specs_map[pkg_name] = MatchSpec(pkg_name, target=target_prec.dist_str())
+
+        pin_overrides = set()
+        for s in ssc.pinned_specs:
+            if s.name in explicit_pool:
+                if s.name not in self.specs_to_add_names:
+                    ssc.specs_map[s.name] = s
+                elif explicit_pool[s.name] & self._get_package_pool(ssc, [s])[s.name]:
+                    ssc.specs_map[s.name] = MatchSpec(s, optional=False)
+                    pin_overrides.add(s.name)
+                else:
+                    log.warn("pinned spec %s conflicts with explicit specs.  "
+                             "Overriding pinned spec.", s)
```

Comment:
Can I ask what is the reason to allow user-specified version to override pinned spec? According to the doc (https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-pkgs.html?highlight=pinned#preventing-packages-from-updating-pinning), the pinned file can prevents packages listed in the pinned file from being updated. This seems like a major behavioral change and I am not sure it is documented on the release notes.

We have multiple use cases that rely on pinned file to prevent accidental version change. Is there a workaround to fallback to the old behavior?

Reply:
I created to https://github.com/conda/conda/issues/9016 to discuss this issue.